src/plugins/auto_batch/auto_batch.cpp

```diff
@@ -0,0 +1,731 @@
+    const DeviceInformation& networkDevice,
+    const std::unordered_map<std::string, InferenceEngine::Parameter>& config,
+    const bool needPerfCounters)
+    : InferenceEngine::ExecutableNetworkThreadSafeDefault(nullptr,
+                                                          std::make_shared<InferenceEngine::ImmediateExecutor>()),
+      _network{networkWithBatch},
+      _networkWithoutBatch{networkWithoutBatch},
+      _config{config},
+      _needPerfCounters{needPerfCounters} {
+    // WA for gcc 4.8 ( fails compilation with member init-list)
+    _device = networkDevice;
+    auto time_out = config.find(CONFIG_KEY(AUTO_BATCH_TIMEOUT));
+    if (time_out != config.end())
+        _timeOut = ParseTimeoutValue(time_out->second.as<std::string>());
+}
+
+AutoBatchExecutableNetwork::~AutoBatchExecutableNetwork() {
+    _terminate = true;
+    for (auto w : _workerRequests) {
+        w->_thread.join();
+    }
+    _workerRequests.clear();
+}
+
+unsigned int AutoBatchExecutableNetwork::ParseTimeoutValue(const std::string& s) {
+    auto val = std::stoi(s);
+    if (val < 0)
+        IE_THROW(ParameterMismatch) << "Value for the " << CONFIG_KEY(AUTO_BATCH_TIMEOUT) << " should be unsigned int";
+    return val;
+}
+
+std::shared_ptr<InferenceEngine::RemoteContext> AutoBatchExecutableNetwork::GetContext() const {
+    return _network->GetContext();
+}
+
+InferenceEngine::IInferRequestInternal::Ptr AutoBatchExecutableNetwork::CreateInferRequestImpl(
+    InferenceEngine::InputsDataMap networkInputs,
+    InferenceEngine::OutputsDataMap networkOutputs) {
+    // todo : guard request creation from another thread/on-the-fly
+    auto num = _numRequestsCreated++;
+    auto batch_id = num % _device.batchForDevice;
+    if (!batch_id) {  // need new request
+        _workerRequests.push_back(std::make_shared<WorkerInferRequest>());
+        auto workerRequestPtr = _workerRequests.back();
+        workerRequestPtr->_inferRequestBatched = {_network->CreateInferRequest(), _network._so};
+        workerRequestPtr->_batchSize = _device.batchForDevice;
+        workerRequestPtr->_completionTasks.resize(workerRequestPtr->_batchSize);
+        workerRequestPtr->_inferRequestBatched->SetCallback(
+            [workerRequestPtr, this](std::exception_ptr exceptionPtr) mutable {
+                if (exceptionPtr)
+                    workerRequestPtr->_exceptionPtr = exceptionPtr;
+                IE_ASSERT(workerRequestPtr->_completionTasks.size() == (size_t)workerRequestPtr->_batchSize);
+                // notify the individual requests on the completion
+                for (int c = 0; c < workerRequestPtr->_batchSize; c++) {
+                    workerRequestPtr->_completionTasks[c]();
+                }
+                // reset the timeout
+                workerRequestPtr->_cond.notify_one();
+            });
+
+        workerRequestPtr->_thread = std::thread([workerRequestPtr, this] {
```

Comment:
It is better to use our executor manager and cached thread to eliminate thread recreation.

Reply:
good point - to minimize the mem fragmentation, will fix in the brushing "PR"